test/tree_diff_tests/compare.py

```diff
@@ -0,0 +1,76 @@
+import copy
+import fnmatch
+import json
+import subprocess
+from dataclasses import dataclass
+from typing import List
+
+
+def compare_trees(a, b):
+    result = subprocess.run(["./tree-diff", a, b], stdout=subprocess.PIPE, check=True)
+    return json.loads(result.stdout)
+
+
+@dataclass
+class File:
+    path: str
+    mode: bool = False
+    uid: bool = False
+    gid: bool = False
+    selinux: bool = False
+    content: bool = False
+
+
+def assert_tree_changes(tree_diff, added=[], deleted=[], modified: List[File] = [], ignored: List[str] = [],
+                        allow_other_changes=False):
```

Comment:
`unittest`'s `assertEqual` does deep comparisons like this on arbitrary data structures. Why are you rolling your own?

More generally: I know this is not using `unittest`, but I wonder why. I think the `test` directory could benefit from a structure that's more consistent with other python projects, making it easier for people to find their way around it. Also, `__main__` is getting quite unwieldy. There's a large ecosystem of test runners and reporters around `unittest` that we could tap into. Do we have a strong reason to deviate from it?

Reply:
tagging @msehnout: Please, can you look up the discussion about test frameworks and osbuild?

Reply:
ok, so it is available in Telegram (it was on += on August 12th) and the outcome is simple: we postponed the discussion, but we have never really had time to continue the it since then...

Reply:
I've added the dicussion to the Berlin meeting agenda.

Reply:
> `unittest`'s `assertEqual` does deep comparisons like this on arbitrary data structures. Why are you rolling your own?

The actual question is: do we want deep equality? Currently, we are not able to make reproducible builds (the stuff with unique machine-id inside /boot/loader/entries). Therefore, it isn't possible to use equality without some exceptions.

Without the tree caching there would be much more "false positives", every tree diff would have differences inside /boot/loader/entries and friends. Should be tree caching even used in these kinds of tests? I think we should maybe turn it off to avoid confusion...